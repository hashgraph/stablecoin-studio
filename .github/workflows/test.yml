name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test