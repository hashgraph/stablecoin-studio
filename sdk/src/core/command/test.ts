/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
