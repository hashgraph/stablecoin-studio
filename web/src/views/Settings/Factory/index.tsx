import { Button, Flex, Heading, SimpleGrid, Text, useDisclosure, VStack } from '@chakra-ui/react';

import {
	Network,
	ChangeFactoryProxyOwnerRequest,
	UpgradeFactoryImplementationRequest,
} from '@hashgraph-dev/stablecoin-npm-sdk';

import { useState } from 'react';
import type { ReactNode } from 'react';
import type { FieldValues } from 'react-hook-form';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import BaseContainer from '../../../components/BaseContainer';
import InputController from '../../../components/Form/InputController';
import ModalNotification from '../../../components/ModalNotification';
import { propertyNotFound } from '../../../constant';

import SDKService from '../../../services/SDKService';
import {
	SELECTED_NETWORK_FACTORY_PROXY_CONFIG,
	IS_FACTORY_PROXY_OWNER,
	walletActions,
	SELECTED_WALLET_ACCOUNT_INFO,
} from '../../../store/slices/walletSlice';
import { handleRequestValidation } from '../../../utils/validationsHelper';
import { RouterManager } from '../../../Router/RouterManager';
import { NamedRoutes } from '../../../Router/NamedRoutes';

const FactorySettings = () => {
	const navigate = useNavigate();
	const dispatch = useDispatch();
	const [awaitingUpdate, setAwaitingUpdate] = useState<boolean>(false);
	const [success, setSuccess] = useState<boolean>();
	const [error, setError] = useState<any>();
	const { isOpen, onOpen, onClose } = useDisclosure();
	const variant = awaitingUpdate ? 'loading' : success ? 'success' : 'error';

	const { t } = useTranslation(['settings', 'global']);

	const factoryProxyConfig = useSelector(SELECTED_NETWORK_FACTORY_PROXY_CONFIG);
	const isFactoryProxyOwner = useSelector(IS_FACTORY_PROXY_OWNER);
	const accountInfo = useSelector(SELECTED_WALLET_ACCOUNT_INFO);

	const [upgradeFactoryImplementationRequest] = useState<UpgradeFactoryImplementationRequest>(
		new UpgradeFactoryImplementationRequest({
			factoryId: '',
			implementationAddress: '',
		}),
	);
	const [changeFactoryProxyOwnerRequest] = useState<ChangeFactoryProxyOwnerRequest>(
		new ChangeFactoryProxyOwnerRequest({
			factoryId: '',
			targetId: '',
		}),
	);

	const form = useForm<FieldValues>({
		mode: 'onChange',
	});

	const { control, getValues } = form;

	const handleUpgradeImplementation = async () => {
		const { updateImplementation } = getValues();
		const factoryId: string = await Network.getFactoryAddress();

		if (factoryId) {
			upgradeFactoryImplementationRequest.factoryId = factoryId;
			upgradeFactoryImplementationRequest.implementationAddress = updateImplementation;

			try {
				onOpen();
				setAwaitingUpdate(true);
				await SDKService.upgradeFactoryImplementation(upgradeFactoryImplementationRequest);
				setError('');
				setAwaitingUpdate(false);
				setSuccess(true);
				dispatch(
					walletActions.setSelectedNetworkFactoryProxyConfig({
						owner: factoryProxyConfig?.owner?.toString(),
						implementationAddress: updateImplementation.toString(),
					}),
				);
			} catch (error: any) {
				setAwaitingUpdate(false);
				console.log(error);
				setError(error?.transactionError?.transactionUrl);
				setSuccess(false);
				setAwaitingUpdate(false);
			}
		}
	};

	const handleChangeOwner = async () => {
		const { updateOwner } = getValues();
		const factoryId: string = await Network.getFactoryAddress();

		if (factoryId) {
			changeFactoryProxyOwnerRequest.factoryId = factoryId;
			changeFactoryProxyOwnerRequest.targetId = updateOwner;

			try {
				onOpen();
				setAwaitingUpdate(true);
				await SDKService.changeFactoryOwner(changeFactoryProxyOwnerRequest);
				setError('');
				setAwaitingUpdate(false);
				setSuccess(true);
				dispatch(
					walletActions.setIsFactoryProxyOwner(
						updateOwner.toString() === accountInfo?.id?.toString(),
					),
				);
				dispatch(
					walletActions.setSelectedNetworkFactoryProxyConfig({
						owner: updateOwner.toString(),
						implementationAddress: factoryProxyConfig?.implementationAddress?.toString(),
					}),
				);

				RouterManager.to(navigate, NamedRoutes.Settings);
			} catch (error: any) {
				setAwaitingUpdate(false);
				console.log(error);
				setError(error?.transactionError?.transactionUrl);
				setSuccess(false);
				setAwaitingUpdate(false);
			}
		}
	};

	const GridItem = ({
		name,
		title,
		label,
		current,
		input,
		button,
	}: {
		name: string;
		title: string;
		label: string;
		current: string;
		input: ReactNode;
		button: ReactNode;
	}) => (
		<Flex direction='column' alignItems={'center'}>
			<VStack w='100%' alignItems={'start'} gap='6px' maxWidth={'500px'} alignSelf='center'>
				<Heading
					data-testid={name + '-title'}
					fontSize='24px'
					fontWeight='600'
					mb={8}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{title}
				</Heading>
				<Text data-testid={name + '-label'} style={{ fontWeight: '600' }}>
					{label}
				</Text>
				<Text
					pl='10px'
					data-testid={name + '-value'}
					color={current === 'Error' ? 'red' : ''}
					mb={5}
				>
					{current}
				</Text>
				{input}
				{button}
			</VStack>
		</Flex>
	);

	return (
		<BaseContainer title={t('settings:title')}>
			{isFactoryProxyOwner && (
				<Flex
					direction='column'
					bg='brand.gray100'
					px={{ base: 4, lg: 14 }}
					pt={{ base: 4, lg: 14 }}
					pb={6}
				>
					<SimpleGrid columns={{ base: 2 }} gap={{ base: 10, lg: 20, md: 40 }} alignItems='center'>
						{GridItem({
							name: 'owner',
							title: t('settings:factory.updateOwner.title'),
							label: t('settings:factory.updateOwner.label'),
							current: factoryProxyConfig?.owner?.toString() ?? '',
							input: (
								<InputController
									control={control}
									rules={{
										required: t('global:validations.required') ?? propertyNotFound,
										validate: {
											validation: (value: string) => {
												changeFactoryProxyOwnerRequest.targetId = value;
												const res = handleRequestValidation(
													changeFactoryProxyOwnerRequest.validate('targetId'),
												);
												return res;
											},
										},
									}}
									name={'updateOwner'}
									placeholder={
										t('settings:factory.updateOwner.inputPlaceholder') ?? propertyNotFound
									}
									isReadOnly={false}
								/>
							),
							button: (
								<Button
									data-testid={`update-update-owner-button`}
									variant='primary'
									onClick={handleChangeOwner}
								>
									{t('settings:factory.updateOwner.buttonText')}
								</Button>
							),
						})}
						{GridItem({
							name: 'address',
							title: t('settings:factory.updateImplementation.title'),
							label: t('settings:factory.updateImplementation.label'),
							current: factoryProxyConfig?.implementationAddress?.toString() ?? '',
							input: (
								<InputController
									control={control}
									rules={{
										required: t('global:validations.required') ?? propertyNotFound,
										validate: {
											validation: (value: string) => {
												upgradeFactoryImplementationRequest.implementationAddress = value;
												const res = handleRequestValidation(
													upgradeFactoryImplementationRequest.validate('implementationAddress'),
												);
												return res;
											},
										},
									}}
									name={'updateImplementation'}
									placeholder={
										t('settings:factory.updateImplementation.inputPlaceholder') ?? propertyNotFound
									}
									isReadOnly={false}
								/>
							),
							button: (
								<Button
									data-testid={`update-implementation-address-button`}
									variant='primary'
									onClick={handleUpgradeImplementation}
								>
									{t('settings:factory.updateImplementation.buttonText')}
								</Button>
							),
						})}
					</SimpleGrid>
				</Flex>
			)}
			<ModalNotification
				variant={variant}
				title={
					awaitingUpdate
						? 'Loading'
						: t('settings:factory.notification.title', {
								result: success ? 'Success' : 'Error',
						  })
				}
				description={
					awaitingUpdate
						? undefined
						: t(`settings:factory.notification.description${success ? 'Success' : 'Error'}`)
				}
				isOpen={isOpen}
				onClose={onClose}
				onClick={onClose}
				errorTransactionUrl={error}
				closeButton={false}
				closeOnOverlayClick={false}
			/>
		</BaseContainer>
	);
};

export default FactorySettings;
