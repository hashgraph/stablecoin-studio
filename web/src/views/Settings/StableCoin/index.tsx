import { Button, Flex, Heading, SimpleGrid, Text, useDisclosure, VStack } from '@chakra-ui/react';

import {
	ChangeProxyOwnerRequest,
	GetTokenManagerListRequest,
	Network,
	UpgradeImplementationRequest,
} from '@hashgraph-dev/stablecoin-npm-sdk';

import { useEffect, useState } from 'react';
import type { ReactNode } from 'react';
import type { FieldValues } from 'react-hook-form';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import AwaitingWalletSignature from '../../../components/AwaitingWalletSignature';
import BaseContainer from '../../../components/BaseContainer';
import InputController from '../../../components/Form/InputController';
import ModalNotification from '../../../components/ModalNotification';
import { propertyNotFound } from '../../../constant';

import { useRefreshCoinInfo } from '../../../hooks/useRefreshCoinInfo';
import SDKService from '../../../services/SDKService';
import {
	SELECTED_WALLET_COIN_PROXY_CONFIG,
	SELECTED_WALLET_COIN,
	IS_PROXY_OWNER,
	walletActions,
	SELECTED_WALLET_ACCOUNT_INFO,
} from '../../../store/slices/walletSlice';
import { handleRequestValidation } from '../../../utils/validationsHelper';
import type { Option } from '../../../components/Form/SelectCreatableController';
import SelectCreatableController from '../../../components/Form/SelectCreatableController';
import { RouterManager } from '../../../Router/RouterManager';
import { NamedRoutes } from '../../../Router/NamedRoutes';

const StableCoinSettings = () => {
	const navigate = useNavigate();
	const dispatch = useDispatch();
	const [awaitingUpdate, setAwaitingUpdate] = useState<boolean>(false);
	const [success, setSuccess] = useState<boolean>();
	const [error, setError] = useState<any>();
	const { isOpen, onOpen, onClose } = useDisclosure();
	const variant = awaitingUpdate ? 'loading' : success ? 'success' : 'error';
	const [optionshederaTokenManagerAddresses, setOptionsHederaTokenManagerAddresses] = useState<
		Option[]
	>([]);
	const [gettingHederaTokenManager, setGettingHederaTokenManager] = useState<boolean>(false);

	const { t } = useTranslation(['settings', 'global']);

	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const proxyConfig = useSelector(SELECTED_WALLET_COIN_PROXY_CONFIG);
	const isProxyOwner = useSelector(IS_PROXY_OWNER);
	const accountInfo = useSelector(SELECTED_WALLET_ACCOUNT_INFO);

	const [upgradeImplementationRequest] = useState<UpgradeImplementationRequest>(
		new UpgradeImplementationRequest({
			tokenId: '',
			implementationAddress: '',
		}),
	);
	const [changeProxyOwnerRequest] = useState<ChangeProxyOwnerRequest>(
		new ChangeProxyOwnerRequest({
			tokenId: '',
			targetId: '',
		}),
	);

	const { isLoading } = useRefreshCoinInfo();

	const form = useForm<FieldValues>({
		mode: 'onChange',
	});

	const optionsHederaTokenManager = async () => {
		setGettingHederaTokenManager(true);

		try {
			const hederaTokenManagerOption: any = await Promise.race([
				SDKService.getHederaTokenManagerList(
					new GetTokenManagerListRequest({
						factoryId: await Network.getFactoryAddress(),
					}),
				),
				new Promise((resolve, reject) => {
					setTimeout(() => {
						reject(
							new Error("TokenManager contracts list couldn't be obtained in a reasonable time."),
						);
					}, 10000);
				}),
			]).catch((e) => {
				console.log(e.message);
				onOpen();
				throw e;
			});

			const AllOptions: any[] = [];
			AllOptions.push({
				value: '',
				label: 'Enter your own HederaTokenManager implementation',
				isDisabled: true,
			});

			const options = hederaTokenManagerOption.map((item: any) => {
				return { label: item.value, value: item.value };
			});

			options.forEach((option: any) => {
				if (option.value.toString() !== proxyConfig?.implementationAddress?.toString())
					AllOptions.push(option);
			});

			setOptionsHederaTokenManagerAddresses(AllOptions.reverse());

			setGettingHederaTokenManager(false);
		} catch (e) {
			setGettingHederaTokenManager(false);
			throw e;
		}
	};

	useEffect(() => {
		optionsHederaTokenManager().catch(console.error);
	}, []);

	useEffect(() => {
		optionsHederaTokenManager().catch(console.error);
	}, [proxyConfig]);

	const { control, getValues } = form;

	const handleUpgradeImplementation = async () => {
		const { updateImplementation } = getValues();
		const implementationAddress = updateImplementation.value;
		if (selectedStableCoin?.tokenId) {
			upgradeImplementationRequest.tokenId = selectedStableCoin.tokenId.toString();
			upgradeImplementationRequest.implementationAddress = implementationAddress;

			try {
				onOpen();
				setAwaitingUpdate(true);
				await SDKService.upgradeImplementation(upgradeImplementationRequest);
				setError('');
				setAwaitingUpdate(false);
				setSuccess(true);
				dispatch(
					walletActions.setSelectedStableCoinProxyConfig({
						owner: proxyConfig?.owner?.toString(),
						implementationAddress: implementationAddress.toString(),
					}),
				);
			} catch (error: any) {
				setAwaitingUpdate(false);
				console.log(error);
				setError(error?.transactionError?.transactionUrl);
				setSuccess(false);
				setAwaitingUpdate(false);
			}
		}
	};

	const handleChangeOwner = async () => {
		const { updateOwner } = getValues();
		if (selectedStableCoin?.tokenId) {
			changeProxyOwnerRequest.tokenId = selectedStableCoin.tokenId.toString();
			changeProxyOwnerRequest.targetId = updateOwner;

			try {
				onOpen();
				setAwaitingUpdate(true);
				await SDKService.changeOwner(changeProxyOwnerRequest);
				setError('');
				setAwaitingUpdate(false);
				setSuccess(true);
				dispatch(
					walletActions.setIsProxyOwner(updateOwner.toString() === accountInfo?.id?.toString()),
				);
				dispatch(
					walletActions.setSelectedStableCoinProxyConfig({
						owner: updateOwner.toString(),
						implementationAddress: proxyConfig?.implementationAddress?.toString(),
					}),
				);

				RouterManager.to(navigate, NamedRoutes.Settings);
			} catch (error: any) {
				setAwaitingUpdate(false);
				console.log(error);
				setError(error?.transactionError?.transactionUrl);
				setSuccess(false);
				setAwaitingUpdate(false);
			}
		}
	};

	const GridItem = ({
		name,
		title,
		label,
		current,
		input,
		button,
	}: {
		name: string;
		title: string;
		label: string;
		current: string;
		input: ReactNode;
		button: ReactNode;
	}) => (
		<Flex direction='column' alignItems={'center'}>
			<VStack w='100%' alignItems={'start'} gap='6px' maxWidth={'500px'} alignSelf='center'>
				<Heading
					data-testid={name + '-title'}
					fontSize='24px'
					fontWeight='600'
					mb={8}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{title}
				</Heading>
				<Text data-testid={name + '-label'} style={{ fontWeight: '600' }}>
					{label}
				</Text>
				<Text
					pl='10px'
					data-testid={name + '-value'}
					color={current === 'Error' ? 'red' : ''}
					mb={5}
				>
					{current}
				</Text>
				{input}
				{button}
			</VStack>
		</Flex>
	);

	return (
		<BaseContainer title={t('settings:title')}>
			{(isLoading || gettingHederaTokenManager) && <AwaitingWalletSignature />}
			{selectedStableCoin &&
				!isLoading &&
				!gettingHederaTokenManager &&
				selectedStableCoin.proxyAdminAddress &&
				isProxyOwner && (
					<Flex
						direction='column'
						bg='brand.gray100'
						px={{ base: 4, lg: 14 }}
						pt={{ base: 4, lg: 14 }}
						pb={6}
					>
						<SimpleGrid
							columns={{ base: 2 }}
							gap={{ base: 10, lg: 20, md: 40 }}
							alignItems='center'
						>
							{GridItem({
								name: 'owner',
								title: t('settings:stableCoin.updateOwner.title'),
								label: t('settings:stableCoin.updateOwner.label'),
								current: proxyConfig?.owner?.toString() ?? '',
								input: (
									<InputController
										control={control}
										rules={{
											required: t('global:validations.required') ?? propertyNotFound,
											validate: {
												validation: (value: string) => {
													changeProxyOwnerRequest.targetId = value;
													const res = handleRequestValidation(
														changeProxyOwnerRequest.validate('targetId'),
													);
													return res;
												},
											},
										}}
										name={'updateOwner'}
										placeholder={
											t('settings:stableCoin.updateOwner.inputPlaceholder') ?? propertyNotFound
										}
										isReadOnly={false}
									/>
								),
								button: (
									<Button
										data-testid={`update-owner-button`}
										variant='primary'
										onClick={handleChangeOwner}
									>
										{t('settings:stableCoin.updateOwner.buttonText')}
									</Button>
								),
							})}
							{GridItem({
								name: 'address',
								title: t('settings:stableCoin.updateImplementation.title'),
								label: t('settings:stableCoin.updateImplementation.label'),
								current: proxyConfig?.implementationAddress?.toString() ?? '',
								input: (
									<SelectCreatableController
										overrideStyles={{
											wrapper: {
												border: '1px',
												borderColor: 'brand.black',
												borderRadius: '8px',
												height: 'min',
											},
											menuList: {
												maxH: '220px',
												overflowY: 'auto',
												bg: 'brand.white',
												boxShadow: 'down-black',
												p: 2,
												zIndex: 99,
											},
											valueSelected: {
												fontSize: '14px',
												fontWeight: '500',
											},
										}}
										addonLeft={true}
										rules={{
											required: t(`global:validations.required`) ?? propertyNotFound,
											validate: {
												validation: (option: any) => {
													if (!option || !option.value) return false;
													upgradeImplementationRequest.implementationAddress =
														option.value as string;
													const res = handleRequestValidation(
														upgradeImplementationRequest.validate('implementationAddress'),
													);
													return res;
												},
											},
										}}
										variant='unstyled'
										name={`updateImplementation`}
										control={control}
										isRequired={true}
										defaultValue={'0'}
										options={[...Object.values(optionshederaTokenManagerAddresses)]}
										placeholder={
											t('settings:stableCoin.updateImplementation.inputPlaceholder') ??
											propertyNotFound
										}
									/>
								),
								button: (
									<Button
										data-testid={`update-implementation-address-button`}
										variant='primary'
										onClick={handleUpgradeImplementation}
									>
										{t('settings:stableCoin.updateImplementation.buttonText')}
									</Button>
								),
							})}
						</SimpleGrid>
					</Flex>
				)}
			<ModalNotification
				variant={variant}
				title={
					awaitingUpdate
						? 'Loading'
						: t('settings:stableCoin.notification.title', {
								result: success ? 'Success' : 'Error',
						  })
				}
				description={
					awaitingUpdate
						? undefined
						: t(`settings:stableCoin.notification.description${success ? 'Success' : 'Error'}`)
				}
				isOpen={isOpen}
				onClose={onClose}
				onClick={onClose}
				errorTransactionUrl={error}
				closeButton={false}
				closeOnOverlayClick={false}
			/>
		</BaseContainer>
	);
};

export default StableCoinSettings;
