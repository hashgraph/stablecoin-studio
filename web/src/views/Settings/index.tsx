import {
	Box,
	Heading,
	Alert,
	AlertDescription,
	AlertIcon,
	CloseButton,
	Flex,
	Link,
	Text,
	Image,
} from '@chakra-ui/react';
import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import BaseContainer from '../../components/BaseContainer';
import type { DirectAccessProps } from '../../components/DirectAccess';
import GridDirectAccess from '../../components/GridDirectAccess';
import type { IAccountToken } from '../../interfaces/IAccountToken';
import type { IExternalToken } from '../../interfaces/IExternalToken';
import {
	SELECTED_WALLET_COIN,
	SELECTED_WALLET_PAIRED_ACCOUNTID,
	IS_PROXY_OWNER,
	IS_FACTORY_PROXY_OWNER,
	IS_PENDING_OWNER,
	IS_ACCEPT_OWNER,
	IS_FACTORY_ACCEPT_OWNER,
	IS_FACTORY_PENDING_OWNER,
	SELECTED_NETWORK_FACTORY_PROXY_CONFIG,
} from '../../store/slices/walletSlice';
import { NamedRoutes } from './../../Router/NamedRoutes';
import { RouterManager } from '../../Router/RouterManager';
import SAFE_BOX from '../../assets/svg/safe-box.svg';

const Settings = () => {
	const { t } = useTranslation(['settings', 'errorPage']);
	const navigate = useNavigate();

	const selectedStableCoin = useSelector(SELECTED_WALLET_COIN);
	const accountId = useSelector(SELECTED_WALLET_PAIRED_ACCOUNTID);
	const isProxyOwner = useSelector(IS_PROXY_OWNER);
	const isPendingOwner = useSelector(IS_PENDING_OWNER);
	const isAcceptOwner = useSelector(IS_ACCEPT_OWNER);
	const isFactoryProxyOwner = useSelector(IS_FACTORY_PROXY_OWNER);
	const isFactoryPendingOwner = useSelector(IS_FACTORY_PENDING_OWNER);
	const isFactoryAcceptOwner = useSelector(IS_FACTORY_ACCEPT_OWNER);
	const selectedNetworkFactoryProxyConfig = useSelector(SELECTED_NETWORK_FACTORY_PROXY_CONFIG);

	const [showAlert, setShowAlert] = useState<boolean>(false);

	const [disabledFeatures, setDisabledFeatures] = useState({
		stableCoin: false,
		factory: false,
	});

	useEffect(() => {
		getAvailableFeatures();
	}, [
		selectedStableCoin,
		isProxyOwner,
		isPendingOwner,
		isAcceptOwner,
		isFactoryProxyOwner,
		isFactoryPendingOwner,
		isFactoryAcceptOwner,
	]);

	const getAvailableFeatures = async () => {
		let isExternalToken = false;

		const tokensAccount = localStorage?.tokensAccount;
		if (tokensAccount) {
			const tokensAccountParsed = JSON.parse(tokensAccount);
			if (tokensAccountParsed) {
				const myAccount = tokensAccountParsed.find(
					(acc: IAccountToken) => acc.id === accountId?.toString(),
				);
				if (myAccount) {
					const externalToken = myAccount?.externalTokens.find(
						(coin: IExternalToken) => coin.id === selectedStableCoin?.tokenId?.toString(),
					);
					if (externalToken) {
						isExternalToken = true;
					}
				}
			}
		}

		const areDisabled = {
			stableCoin:
				!selectedStableCoin ||
				(selectedStableCoin &&
					((isExternalToken && !isProxyOwner && !isAcceptOwner) ||
						(!isProxyOwner && !isAcceptOwner))),
			factory: !isFactoryProxyOwner && !isFactoryAcceptOwner,
		};
		setDisabledFeatures(areDisabled);
	};

	if (!selectedStableCoin && disabledFeatures?.factory) {
		return (
			<Flex
				data-testid='stable-coin-not-selected-container'
				bgColor='white'
				w='100%'
				h='100%'
				justifyContent='center'
				alignItems='center'
				flexDirection='column'
			>
				<Image
					data-testid='stable-coin-not-selected-logo'
					src={SAFE_BOX}
					alt='Safe box'
					w='140px'
					h='140px'
					mb='40px'
				/>
				<Text
					data-testid='stable-coin-not-selected-title'
					fontSize='22px'
					fontWeight='700'
					lineHeight='16px'
					mb='16px'
				>
					{t('errorPage:stableCoinNotSelected.title')}
				</Text>
				<Text
					data-testid='stable-coin-not-selected-description'
					fontSize='16px'
					fontWeight='500'
					lineHeight='16px'
					maxW='500px'
					textAlign='center'
					mb='2px'
				>
					{t('errorPage:stableCoinNotSelected.description')}
				</Text>
				<Text fontSize='16px' fontWeight='500' lineHeight='16px' maxW='500px' textAlign='center'>
					{t('errorPage:stableCoinNotSelected.description2')}
				</Text>
			</Flex>
		);
	}

	const handleFactoryClick = async () => {
		if (!selectedNetworkFactoryProxyConfig) {
			setShowAlert(true);
		} else {
			RouterManager.to(navigate, NamedRoutes.FactorySettings);
		}
	};

	const directAccesses: DirectAccessProps[] = [
		{
			icon: 'CoinVertical',
			route: NamedRoutes.StableCoinSettings,
			title: t('stableCoinSettings'),
			isDisabled: disabledFeatures?.stableCoin,
		},
		{
			icon: 'Factory',
			route: NamedRoutes.FactorySettings,
			title: t('factorySettings'),
			isDisabled: disabledFeatures?.factory,
			customHandleClick: handleFactoryClick,
		},
	];

	const filteredDirectAccesses = directAccesses.filter((access) => !access.isDisabled);

	return (
		<BaseContainer title={t('title')}>
			<Box p={{ base: 4, md: '128px' }}>
				<Heading fontSize='20px' fontWeight='600' mb={14} data-testid='subtitle'>
					{t('settings:subtitle')}
				</Heading>
				{showAlert && (
					<Alert status='warning' justifyContent='center' mb={14}>
						<Flex width='auto'>
							<AlertIcon />
							<Box>
								<AlertDescription>
									<p>{t('settings:alertNoEnv')}</p>
									<Link
										textDecoration='underline'
										isExternal={true}
										href='https://github.com/hashgraph/hedera-accelerator-stablecoin/tree/main/web#env-vars'
									>
										{t('settings:alertNoEnvLink')}
									</Link>
								</AlertDescription>
							</Box>
						</Flex>
						<CloseButton
							alignSelf='flex-start'
							position='relative'
							right={-1}
							top={-1}
							onClick={() => setShowAlert(false)}
						/>
					</Alert>
				)}
				<GridDirectAccess directAccesses={filteredDirectAccesses} />
			</Box>
		</BaseContainer>
	);
};

export default Settings;
