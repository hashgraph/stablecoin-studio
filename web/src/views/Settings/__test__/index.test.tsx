import Settings from '../';
import { render } from '../../../test/index';
import translations from '../../../translations/en/settings.json';

describe(`<${Settings.name} />`, () => {
	beforeEach(() => {});

	test('should render correctly', () => {
		localStorage.setItem(
			'tokensAccount',
			JSON.stringify([
				{
					id: 'id',
					externalTokens: [
						{
							id: '0.0.12345',
							symbol: 'symbol',
						},
					],
				},
			]),
		);
		const component = render(<Settings />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<Settings />);
		const header = component.getByTestId('base-container-heading');

		expect(header).toHaveTextContent(translations.title);
	});

	test('should not allow to select stable coin settings', async () => {
		const component = render(<Settings />);
		const stablecoinSettingsButton = component.getByTestId('direct-access-stableCoinSettings');

		expect(stablecoinSettingsButton).toBeDisabled();
	});

	test('should not allow to select factory settings', async () => {
		const component = render(<Settings />);
		const factorySettingsButton = component.getByTestId('direct-access-factorySettings');

		expect(factorySettingsButton).toBeDisabled();
	});
});
