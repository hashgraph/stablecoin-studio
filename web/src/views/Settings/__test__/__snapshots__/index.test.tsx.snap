// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-19t9j1"
  >
    <h2
      class="chakra-heading css-bqsusc"
      data-testid="base-container-heading"
    >
      Settings
    </h2>
    <div
      class="css-1b6lda7"
      data-testid="base-container-children"
    >
      <div
        class="css-src872"
      >
        <h2
          class="chakra-heading css-qltjox"
          data-testid="subtitle"
        >
          Choose an option
        </h2>
        <div
          class="css-nxcc5s"
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
